### ISSUE TITLE
Unexpected shallow routing when navigating through history.

### ISSUE THREAD

**Issue by eirikurn** [zeit/next.js#7395]
# Bug report

## Describe the bug

When going back through history with a dynamic page, NextJS will use newer props to render shallow history entries.

## To Reproduce

Say I have a dynamic route that handles `/article/{id}` and `/article/{id}/comments`. Given the same id, these URLs use the same data with a tab widget to switch between the two views and URLs. So we can use shallow routing to switch between the tabs without scrolling or refetching data.

The issue manifests with this kind of navigation:

1. Go to `/article/hello-world`. This calls `getInitialProps` to fetch article "hello-world".
2. Switch to comments tab with a shallow link to `/article/hello-world/comments`.
3. Click a link to `/article/nextjs-rocks`. This calls `getInitialProps` to fetch article "nextjs-rocks".
4. Click browser back button. Url changes to `/article/hello-world/comments`, but the page shows comments for "nextjs-rocks" article

[Try it here](https://nextjs-shallow-issue-eirikurn.aranja1.now.sh/). [Code](https://github.com/eirikurn/nextjs-shallow-issue)

## Expected behavior

See comments for "hello-world" article.

This definitely needs a deeper discussion, but I'd like to group shallow history entries with the history entry that initially called `getInitialProps`. So when someone navigates back from `/article/nextjs-rocks` to the shallow `/article/hello-world/comments` entry, it calls `getInitialProps` to fetch data. When they go further back to `/article/hello-world`, that could perhaps be shallow.

## System information
 - Version of Next.js: 8.0.4

## Context

I encountered this issue with a dynamic page that pulls content from a CMS. It broke when I created a component that cleans up trailing slashes in the url, by calling Router.replace with shallow: true. When it cleaned the url, I could not go back to that route, since it would use props with a different page from the CMS.

**Comment by EyMaddis** [comment]
@eirikurn Did you find a way around that?

**Comment by eirikurn** [comment]
No. I had to remove this logic client side. So We only clean urls with server side redirects currently.

**Comment by EyMaddis** [comment]
I was able to work around it, but this only works in my specific setup:

I use apollo-ssr with getDataFromTree instead of using `getInitialProps` for data loading, which does not run into [the problem that `getIntialProps` won't be called again](https://github.com/zeit/next.js/issues/3065)...

```jsx
function Page({ id: idFromNextJs }) {
  const idToLoadDataFor = (process.browser ? Router.query?.id  : idFromNextJs) || idFromNextJs
  const { data } = useQuery(...., { variables: { id: idToLoadDataFor } })
...
```

Far from perfect, but works.

**Comment by fabb** [comment]
We could save a lot of server load if we could avoid SSRs in this case. Any ideas for workarounds?

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
